File: src/zmq_experiment/pub/publisher.py
import zmq
import zmq.asyncio

from ..audio_packet import AudioPacket


class Publisher:
    _context: zmq.asyncio.Context
    _socket: zmq.asyncio.Socket

    def __init__(self):
        self._context = zmq.asyncio.Context()
        self._socket = self._context.socket(zmq.PUSH)
        self._socket.bind("tcp://*:5555")

    async def publish(self, packet: AudioPacket):
        await self._socket.send(packet.serialize())

File: src/cpp/src/publisher.cpp
#include "Publisher.hpp"

#include <filesystem>
#include <ostream>

Publisher::Publisher() : context(1), socket(context, zmq::socket_type::push)
{
    socket.bind("tcp://*:5555");
}

Publisher::~Publisher()
{
    socket.close();
}

void Publisher::publish(const AudioPacket &packet)
{
    auto serialized = packet.serialize();
    zmq::message_t zmq_message(serialized.data(), serialized.size());
    socket.send(zmq_message, zmq::send_flags::none);
}
